### PR TITLE
Cigar P support plus bug fixes for I/D and D/I combinations

### LINKED ISSUE #734: Possible error in samtools tview
Hi,

In the process of troubleshooting my own pileup generator I have noticed that samtools sometimes generates pileups that imply nonexistent read sequences. For example, using nanopore reads ([data](https://github.com/kishwarshafin/deePore/tree/master/data)), the following pileup is generated:

```
samtools tview -p chr3:79005 -d T /hive/users/rlorigro/data/NA12878.np.chr3.100kb.0.bam /hive/users/rlorigro/data/hg19.chr3.9mb.fa 
      79011      79021      79031       79041        79051      79061           
tgttgagattacag*gcgtgag*ccactgcgcctgg*c*taattttt***gtatttttagtag*gaacagagtttcac*c
.............. ....... ............. . ........   ............. .............. .
,,,,,,,,,,,,,,*,*,c,,,*,,,,,,**,,,,,*,*,,,,,*,,***,,,,,,,,,,c,,g,,,,,,,,,,,,****
,,,,,,,,,,,,,,*****,,,*,,,,c,,,,,,,**,t,,,,,,*,gcg,,,,,*,,,,,,,**,,,,,,,,,,,,,c,
 ,,,,,,,,,,t,,g,,,,,,ttt,,a,,,,t,***c,***,c,,,,***,,,,,,,,,,,,,*,,,,,,,,**,,,,*,
```

Looking closely at this region:

```
gcctgg*c*taatt
...... . .....
*,,,,,*,*,,,,,
,,,,,**,t,,,,,
,t,***c,***,c,
```

The last alignment implies the sequence: "GTCCCACTTT"

Whereas my own interpretation of the cigar string would imply this sequence: "GTCACACTTT":

```
GTTGAGATTACAG_GCGTGAG_CCACTGCGCCTGG_C_TAATTTTT___GTATTTTTA
GTTGAGATTACAG_G*GCGAG_CCACTG**CCTGG_C_TAATT*TT___GTATTTTTA
GTTGAGATTACAG_****GAG_CCACCGCGCCTG*_CTTAATTT*TGCGGTATT*TTA
GTTGAGATTATAGGGCGTGATTTCAATGCGTC***AC_**ACTTTT___GTATTTTTA
```

Searching the SAM file with grep yields two different reads, and the sequence suggested by samtools tview doesn't exist in the correct location for it to have been mapped here:

```
>grep "GTCCCACTTT" NA12878.np.chr3.100kb.0.sam 
210fdd9b-6701-4977-b559-6b6fc15039aa_Basecall_Alignment_template	16	chr3	59965	0 ... TTGTAAGTCCCACTTTAAGAT ...
```

In other words, the sequence from samtools doesn't actually exist in the query region

```
 >grep "GTCACACTTT" NA12878.np.chr3.100kb.0.sam 
55d0825e-4caf-4c1d-8d9b-7b6416de9874_Basecall_Alignment_template	16	chr3	79006	0 ... CAATGCGTCACACTTTTGTAT ...
```

Whereas this one does.

How is it that samtools substitutes a C for an A here? Is this something related to left alignment of indels? I am using v1.5

### LINKED ISSUE #842: mpileup gives some insertions with "=" at the end
Using samtools 1.7 on Linux 2.6.32-504.el6.x86_64 with gcc 4.4.7.

Mpileup gives "=" at the end of some insertions in the pileup file (this character is count in the length of the insertion):

```
[]$ samtools mpileup -f ucsc.hg19.fasta -r chr11:62397115-62397115 -L 1000000 -d 1000000 --min-BQ 20 --min-MQ 20 -B -R input.bam > tmp.pileup
[mpileup] 1 samples in 1 input files

[]$ cat tmp.pileup
chr11	62397115	G	517	.,-1c,-1c,-1c,tt.*+4CCGA$*+4CCGA$.-1C*.......*+3CC=$..C..C.*+9CCGATCGC=$C.*+5CCGA=$*+4CCGA$.*+3CC=$*+6CCGACA$.*+10CCGATCGCCA$*+8CCGATTCA$..*+6CCGATA$.....t,-1ct,tt,t,,-1c,,tttttt,-1c,,tt,,t,-1ct,,,t,,,-1ct,,-1c,,,t,,-1c,,,t,-1c,,-1ct,-1c,-1ct,+1c,,-1c,,,,-1ct,,,tt,-1c,,-1c,,,,,-1c,-1c,t,-1c,,,,-1c,,,+1c,t,-1c,-1c,tt,t,,,,,-1c,t,,t,,,-1c,,,,t,-1c,,-1cttt,,,,,-1ct,,-1c,-1c,tttt,,,,-1c,tta,t,-1c,-1c,,t,-1c,-1c,,,-1c,-1c,,-1ct,-1ct,t,,,-1c,,,-1ct,t,-1c,-1c,-1c,,,,,-1ctt,-1ct,-1c,-1ctt,,,t,-1c,,,t,,,,-1ct,tt,,,,,.-1C,,,-2cct,,,t,,.....-1C...........-1C.......+1C.-1C..........-1C..-1C.............-1C...............................-1C.......-1C.-1C..............................................+1C.....-1C.-1C.......................,,,-1c,,,+1c,-1c,,-1c,-1c,,,-1c,-1c,,,,,,-1c,,-1c,,-1c,-1c,,,,,,,,,,,,,,,,,,,,,,,,,,+2ct,,,,,,...,,...-1C.....-1C.,....+1C,+1ct.,,-1ct,-1ctt,,^ht	<:5565:587:58657:<67755766997::656:7;9766656657:7=586587<85:589;6<8;87588755<7=<55<5875;=777885<57:855=:5<578:<8:65:675=659>9557>5<:=5:75=:;<75:;=7:<==<=?57;>555;5<587:>555=;575885=5;;A7:5=5995<575=785=755658:=8<77855975=;;87;87855<5;5>555555=7=58=99;=7=767<5877<88589:<857===888<<5988:9=75<59<988<87<<<8878<8<99;;8<<=<<;8<889<<9=5<:<<8998985<97889;=;;8998<<<987><<8=88<899<8@89<8<:989<=<98<<>895>8<78<79758<=8888;8;;98<:898<8:879667;87<5677:7;<<>>788:99599;8=;78=9<679;:5757=55<6=68<5=8<8:6B==89<<99899;;887=67659:67
```

Here for example it gives an insertion `+3CC=`.

Here is the 23rd read (should contain this insertion):

```
[]$ samtools view input.bam chr11:62397115-62397115 | head -n23 | tail -n1
H8VG5:01411:13177	0	chr11	62397016	70	71M1D25M1I10M1I21M1I13M	*	0	0	GGCCTTCACACCCTTCTGCCACAGCTCCATTCTCTAAACCTTACCAAAGCGCTGGGAGCCAGCGAAGAAGGCCTGGCCAGGACAAAGGGGCGTTCCGATGCCCCCAGAACCGCTGTCTCAGCCCATCAGCCAGTCGCCATGTG	=9<8<9=>>>>>8>9@===9=>?>=?>9>?9>>><<<6<8<8<<9<<7=<<<<<<4;<<188;;;5:;5:/716<8=8<<7553:(555'3594;-33666;==,::;7<26699::;988::288:8..).555:;7@@<<:	ZP:B:f,0.00572064,0.00442349,8.1542e-06	ZG:i:247	ZB:i:30	ZC:B:i,247,246,1,0	ZA:i:143	ZM:B:s,258,-8,260,-8,-16,236,-36,246,274,-4,-14,2,272,14,218,248,238,494,48,244,242,276,270,260,248,462,4,0,458,-4,12,4,508,30,254,36,0,238,240,28,26,22,44,36,240,12,742,28,460,244,18,-22,236,6,232,-14,30,42,0,4,472,216,-18,228,16,244,-2,258,-20,16,246,22,230,492,0,2,254,4,26,46,446,234,0,24,224,274,12,-2,250,-10,8,642,468,40,26,14,476,270,518,-10,36,710,28,236,36,210,4,210,16,-8,266,18,246,-2,730,220,12,46,198,20,10,78,-6,46,450,156,310,318,30,30,-8,276,-4,492,2,276,22,30,12,-14,508,510,514,-6,188,4,478,18,18,462,102,260,-4,494,-30,220,346,634,986,154,76,40,96,224,400,166,420,206,42,58,-58,52,228,36,106,-16,258,128,196,166,76,1000,162,184,-34,224,-18,392,542,38,298,196,196,150,56,284,208,18,234,30,228,224,-6,64,150,180,638,246,196,280,78,238,52,138,212,56,40,44,-24,-4,416,226,196,6,298,16,196,246,28,62,422,68,26,146,-20,-6,254,56,80,-2,250,88,244,36,172,128,254,228,218,46,66,20,216,220,50,400,6,42,0,244,8,182	ZF:i:25	RG:Z:H8VG5.IonXpress_108	PG:Z:tmap	MD:Z:71^C36G32	NM:i:5	AS:i:108	XM:i:141	XA:Z:map4-1	XS:i:9	XZ:i:0
```

### PR DESCRIPTION
**Pull request by jkbonfield**
Resolves #331 (but renamed to match @jmarshall's rename of my cigar-P htslib PR).
Fixes #842, fixes #734, fixes samtools/htslib#59. 

- CIGAR P is now used to keep insertions in alignment; alters mpileup output and makes tview align correctly.
- Mpileup: D following I is now reported correctly, instead of silently hiding the deletion string.
- Mpileup: I following D now reports the correct bases for the insertion, instead of 1 base further on.

### CODE REVIEW

**Comment by jkbonfield** [comment]
Checks do pass, but are oddly not checking the correct htslib branch again.  I could have sworn I fixed that!  The travis failure is due to compiling against develop branch.

**Comment by daviesrob** [comment]
Rebased with added `--reverse-del` support.

File: bam_plcmd.c [hunk 1]
```diff
@@ -84,15 +86,22 @@ static inline void pileup_seq(FILE *fp, const bam_pileup1_t *p, int pos, int ref
         }
         putc(c, fp);
     } else putc(p->is_refskip? (bam_is_rev(p->b)? '<' : '>') : ((bam_is_rev(p->b) && rev_del) ? '#' : '*'), fp);
+    int del_len = -p->indel;
     if (p->indel > 0) {
-        putc('+', fp); printw(p->indel, fp);
-        for (j = 1; j <= p->indel; ++j) {
-            int c = seq_nt16_str[bam_seqi(bam_get_seq(p->b), p->qpos + j)];
-            putc(bam_is_rev(p->b)? tolower(c) : toupper(c), fp);
+        int len = bam_plp_insertion(p, ks, &del_len);
```
Comment (valeriuo) [inline bam_plcmd.c]: A check (or assertion) for allocation errors inside `bam_plp_insertion` could be added here.

File: bam_tview.c [hunk 2]
```diff
@@ -211,7 +212,8 @@ int tv_pl_func(uint32_t tid, uint32_t pos, int n, const bam_pileup1_t *pl, void
         // calculate maximum insert
         for (i = 0; i < n; ++i) {
             const bam_pileup1_t *p = pl + i;
-            if (p->indel > 0 && max_ins < p->indel) max_ins = p->indel;
+            int len = bam_plp_insertion(p, &ks, NULL);
```
Comment (valeriuo) [inline bam_tview.c]: And also here.

**Comment by valeriuo** [comment]
Merged into develop

**Pull request closed without merging**